contrib/cln/examples/pi.1
.TH PI 1
.SH NAME
pi \- compute decimal Archimedes' constant Pi to arbitrary accuracy.
.SH SYNOPSIS
.B pi
.I "[digits]"
.br
.SH "DESCRIPTION"
The
.BR pi 
command prints 100 decimal digits of Archimedes' constant pi or
a number of digits specified by an integer parameter on the command
line.
.TP
.B \-\-help
Show summary of options.
.TP
.B \-\-version
Show version of program and exit.
.TP
.B \-\-bibliography
Output recommended readings and exit.
.SH "SEE ALSO"
The CLN library is documented fully by
.IR "cln" ,
available via the Info system.
.SH AUTHOR
CLN was written by Bruno Haible <haible@gnu.org> and is 
maintained by Richard Kreckel <kreckel@ginac.de>.


NEWS
 # -*- org -*-
# This file is better edited using Emacs

#+TITLE: Feel++ News
#+AUTHOR: Christophe Prud'homme
#+DATE: <2011-02-26 Sat>
#+EMAIL: christophe.prudhomme@ujf-grenoble.fr
#+OPTIONS: toc:t
#+LATEX_class: cparticle


* New in 0.92.0

** Programming environment

 - support for Boost 1.49 and 1.50
 - imported for Eigen 3.1 in contrib/eigen
 - port to gcc4.7, icc 12.1 and clang 3.1
 - added cmake macro feelpp_add application, see http://code.google.com/p/feelpp/issues/detail?id=11&can=1
 - support Gmsh from both executable and library(preferred and more powerful)
   (Fixes issue 43)
 - support for PETSc 3.3

*** Marseille Mesocenter
 - port to Marseille mesocenter supercomputer (thx to CEMRACS'12)

*** TGCC/Curie
 - port on the TGCC
   Fixes issue 35

*** Macport
 - added scotch Portfile from macport and use esmumps variant for mumps support
 - added PETSc 3.3-p1 support and use suitesparse as well as scotch directly
   from macport instead of downloading/compiling them

** Bugs

 - fixed issue 1: http://code.google.com/p/feelpp/issues/detail?id=1&can=1
 - fixed issue 2: http://code.google.com/p/feelpp/issues/detail?id=2&can=1
 - fixed issue 6: http://code.google.com/p/feelpp/issues/detail?id=6&can=1
 - fixed issue 9: http://code.google.com/p/feelpp/issues/detail?id=9&can=1

** Library

*** feelcore
 - improved benchmarking framework and in particular the handling of the
   statistics provided by the benchmark with views per key and flat views. In
   the last case various files are automatically generated for each benchmark

*** feelalg
 - added eigen(3.1) backend with support for dense and sparse eigen matrices
   Fixes issue 36

*** feelmesh
 - added 2D and 3D mesh adaptation using Gmsh

*** feeldiscr
 - seamless parallelisation of the FunctionSpace data structure in the case of
   product of function spaces
 - lift operator merged with projection operator

*** feelfilters
 - support for mesh generation directly from Gmsh library
 - support for element splitting (Fixes issue 40)

** Documentation



* New in 0.91.1

** Maintainance release

 - overall cleanup in C++ preprocessor macros
 - cleanup in cmake
 - fix some compilation issues and added more stringent compiler flags
 - port to boost1.49
 - Fix issues 1, 2 and 9 (see Google Code, http://code.google.com/p/feelpp/issues/list)

* New in 0.91.0

** Main NEWS

 - Feel++ is now seamlessly parallel
 - Feel++ supports seamlessly interpolation in its variational formulation
   language

** Main contributors for this release

Hopefully nobody has been forgotten

 - Abdoulaye Samake
 - Baptiste Morin
 - Christophe Prud'homme
 - Rachida Chakir
 - Stephane Veys
 - Vincent Chabannes
 - Vincent Doyeux

** Programming Environment

*** Developer changes

 - cmake: overall cleanup of cmake feel++ variables which are now prefixed
   FEELPP
 - cmake: in-source builds forbidden
 - cmake: guards againts invalid CMAKE_BUILD_TYPE
 - cmake: added sse, altivec and neon optimizations
 - cmake: fix bug in make_directory
 - cmake: even more minimal configuration in doc/tutorial
 - fix compilation in APPLE and s390 platforms
 - added macport port for feel++ with support for snow leopard and lion

** Contrib

*** Eigen3

 - Eigen code has been updated to 3.0.5

** Libraries

*** feelcore

 - added a config file with extension .cfg based on Boost.program_options in
   Application classes
 - provide systemConfigRepository() and localConfigRepository() in Environment
   class which return respectively the local and system location of the config
   file

*** feelalg

 - merge the two implementations of aitken relaxation method and provide a
   simplified interface
 - added interface nlSolve interface to backend similar to solve allowing to
   order-free and optional parameters
 - added interface to external solver libraries via petsc such as mumps or
   umfpack. This is configurable via the config file (.cfg)

*** feelpoly

 - various fixes in geometric mapping (and inverse) and polynomialset to handle
   objects of topoligical dimension d-1 living in dimension d space.

*** feeldiscr

 - added implementation of $L_2$ and $H_1$ projector which renders rather
   seamless these projection.
 - fix copy/paste bug in bdf extrapolation, the coefficient were divided by dt
 - support for interpolating functions living on meshes of different
   topoligical dimension
 - added mesh elements and faces extraction using createSubmesh() based on mesh
   filters
 - added lift and trace operator, interface might change
 - BDF:
   - new bdf() free function to generate a new bdf shared_ptr object and
     configure it
#+begin_SRC:
     // vm is the variable map provided by program options
     // note that by default most option get their value from the command line
     // or config files
     auto mybdf = bdf( _space=Xh, _vm=this->vm(), _order=2, _time_step=0.1 );
#+end_SRC:
   - changed options for bdf: now use . to separate  bdf from rest of option in
     order to be able to create section in config files
   - can pass a prefix to handle several bdf in application

*** feelfilters

 - GeoTool:
   - update geotool to support markers for 1D meshes
   - added a new Cube object
   - added a pie object
 - added mshconvert() which allows to generate a gmsh mesh out of mesh formats
   supported by gmsh

*** feelvf

 - support for passing shared_ptr<> to the differential operators and to on()
 - support in the language for test and trial functions not being defined on the
   same mesh
 - Language:
   - added sym() and antisym() to compute the symmetric and antisymmetric parts of
     a matrix expression
   - added access to expression components via () operator
   - allow integration of test and/or trial functions which are defined  on different
     meshes as well as integrating on a different mesh.

** Documentation

*** Manual

 - reorganized completely and updated/corrected the manual
 - updated doxygen documentation


* New in 0.90.0

** Programming Environment


contrib/cln/examples/pi.cc
// Compute decimal Archimedes' constant Pi to arbitrary accuracy.

#include <cln/output.h>
#include <cln/real.h>
#include <cln/real_io.h>
#include <cln/version.h>
#include <cctype>
#include <cstdlib>
#include <cstring>
#include <sstream>

using namespace std;
using namespace cln;

static void
usage (ostream &os)
{
	os << "Usage: pi [digits]\n";
	os << "Compute decimal Archimedes' constant Pi to arbitrary accuracy.\n\n";
	os << "      --help                 display this help and exit\n";
	os << "      --version              output version information and exit\n";
	os << "      --bibliography         output recommended readings and exit\n";
}

int
main (int argc, char * argv[])
{
	long digits = 100;
	if (argc > 1) {
		if (argc == 2 && !strcmp(argv[1],"--help")) {
			usage(cout);
			return 0;
		}
		if (argc == 2 && !strcmp(argv[1],"--version")) {
			cout << "pi (CLN "
			     << version_major << "." << version_minor << "." << version_patchlevel
			     << ")" << endl;
			cout << "Written by Bruno Haible and Richard B. Kreckel." << endl;
			cout << endl;
			cout << "Copyright (C) 1988-2008 Bruno Haible, 2000-2009 Richard B. Kreckel." << endl;
			cout << "This is free software; see the source for copying conditions.  There is NO" << endl;
			cout << "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE." << endl;
			cout << endl;
			return 0;
		}
		if (argc == 2 && !strcmp(argv[1],"--bibliography")) {
			cout << "Recommended readings:\n";
			cout << "\"Pi\", by Joerg Arndt and Christoph Haenel (1999)\n";
			cout << "\"Pi: A Source Book\" by Lennart Berggren, Jonathan Borwein, Peter Borwein (1997)" << endl;
			return 0;
		}
		if (argc == 2 && isdigit(argv[1][0])) {
			digits = atol(argv[1]);
		} else {
			usage(cerr);
			return 1;
		}
	}

	if (digits) {
		cl_F p = pi(float_format(digits));
		// make CLN believe this number has default_float_format to suppress
		// exponent marker which would be quite boring for 3.1416...
		cl_print_flags cpf;
		cpf.default_float_format = float_format(p);
		// We cannot print directly because people get dazed and confused
		// when they see gratuitous digits (Debian Bug #286266.) And we
		// must not "fix" the output routine because print-read consistency
		// is at stake. As a workaround, print into a buffer so we can chop
		// off characters from its end.
		stringstream buf;
		print_real(buf, cpf, p);
		istreambuf_iterator<char> i = buf.rdbuf();
		while (--digits+2>0)
			cout << *(i++);
	}
	cout << endl;

	return 0;
}
